aws/client.go

```diff
@@ -0,0 +1,505 @@
+	return []awstypes.TagSpecification{
+		{
+			ResourceType: resourceType,
+			Tags: []awstypes.Tag{
+				{
+					Key:   aws.String(coreInstanceTag),
+					Value: aws.String(c.prefix),
+				},
+			},
+		},
+	}
+}
+
+func (c *DefaultClient) InstanceState(ctx context.Context, instanceID string) (string, error) {
+	describeInstanceStatus := &ec2.DescribeInstanceStatusInput{
+		IncludeAllInstances: aws.Bool(true),
+		InstanceIds:         []string{instanceID},
+	}
+
+	instanceStatus, err := c.client.DescribeInstanceStatus(ctx, describeInstanceStatus)
+	if err != nil {
+		return "", err
+	}
+
+	return string(instanceStatus.InstanceStatuses[0].InstanceState.Name), nil
+}
+
+func (c *DefaultClient) FindMatchingAMI(ctx context.Context, region, version string) (string, error) {
+	containerIndex, err := index.NewContainer()
+	if err != nil {
+		return "", err
+	}
+
+	if version != "" {
+		coreImageTag, err := containerIndex.Match(ctx, version)
+		if err != nil {
+			return "", err
+		}
+		version = coreImageTag
+	} else {
+		// If no specific core instance version has been provided, use the latest published image.
+		latest, err := containerIndex.Last(ctx)
+		if err != nil {
+			return "", err
+		}
+		version = latest
+	}
+
+	awsIndex, err := index.NewAWS()
+	if err != nil {
+		return "", err
+	}
+
+	imageID, err := awsIndex.Match(ctx, region, version)
+	if err != nil {
+		return "", err
+	}
+
+	return imageID, nil
+}
+
+func (c *DefaultClient) EnsureAndAssociateElasticIPv4Address(ctx context.Context, instanceID, elasticIPv4AddressPool, elasticIPv4Address string) (string, error) {
+	var allocateAddressInput ec2.AllocateAddressInput
+
+	if elasticIPv4Address != "" {
+		allocateAddressInput.Address = &elasticIPv4Address
+	}
+
+	if elasticIPv4AddressPool != "" {
+		allocateAddressInput.CustomerOwnedIpv4Pool = &elasticIPv4AddressPool
+	}
+
+	ipv4Allocation, err := c.client.AllocateAddress(ctx, &allocateAddressInput)
+	if err != nil {
+		return "", err
+	}
+
+	associateAddressInput := &ec2.AssociateAddressInput{
+		AllocationId:       ipv4Allocation.AllocationId,
+		AllowReassociation: aws.Bool(true),
+		InstanceId:         aws.String(instanceID),
+	}
+
+	_, err = c.client.AssociateAddress(ctx, associateAddressInput)
+	if err != nil {
+		return "", err
+	}
+
+	return *ipv4Allocation.PublicIp, nil
+}
+
+func (c *DefaultClient) EnsureSubnet(ctx context.Context, subNetID string) (string, error) {
+	var describeSubnetInput ec2.DescribeSubnetsInput
+
+	// find the default subnet
+	if subNetID == "" {
+		describeSubnetInput = ec2.DescribeSubnetsInput{
+			Filters: []awstypes.Filter{
+				{
+					Name:   aws.String("defaultForAz"),
+					Values: []string{"true"},
+				},
+			},
+		}
+	} else {
+		describeSubnetInput = ec2.DescribeSubnetsInput{
+			Filters: []awstypes.Filter{
+				{
+					Name:   aws.String("subnet-id"),
+					Values: []string{subNetID},
+				},
+			},
+		}
+	}
```

Comment:
```suggestion
        describeSubnetInput = ec2.DescribeSubnetsInput{
	        Filters: []awstypes.Filter{
		        {
			        Name:   aws.String("subnet-id"),
			        Values: []string{subNetID},
		        },
	        },
        }
	if subNetID == "" {
		describeSubnetInput = ec2.DescribeSubnetsInput{
			Filters: []awstypes.Filter{
				{
					Name:   aws.String("defaultForAz"),
					Values: []string{"true"},
				},
			},
		}
	} 	
```

just avoiding nest, but not critical :)